Complete rewrite to python 3 and general cleanup

cgahr opened the pull request:
A while ago, I rewrote journal2ebook from python 2 to python 3 since I had problems running it. 

In the end, as an exercise, I completely rewrote and cleaned up the codebase, removed duplication and (IMO) generally improved the code.

In now uses ruff and mypy for code linting, `pyproject.toml` for configuration and installation and generally satisfies the newest python standards. In principle it is also ready for publication on pypi.


I added a .gitignore and removed all build artifacts, they don't really belong in the repo. For this reason, the diff is bigger then it actually needs to be. To make reviewing this commit easier, it might be easier to review the first few commits individually!

This PR is a response to #34.

adasilva commented on README.md:
```diff
@@ -13,69 +15,16 @@ This program should work cross-platform (Linux, Mac, and Windows) though it has
-* ghostscript - to convert pdf to png
-
-  (http://sourceforge.net/projects/ghostscript/)
+Requirements
```
Users may have to install tkinter system package, https://www.geeksforgeeks.org/how-to-install-tkinter-on-linux/

adasilva reviewed (changes requested):
Looks great! Thanks for the refactor to python3 and general cleanup. I have a few requested changes.

adasilva commented:
Profile saving is not working for me.

cgahr commented on pyproject.toml:
```diff
@@ -0,0 +1,145 @@
+    "appdirs",
+    "image",
+    "pdf2image"
+]
```
Nice catch! 

For tk, isn't tkinter part of core python? https://docs.python.org/3/library/tkinter.html

adasilva replied:
ooh looks like it is.

cgahr commented on pyproject.toml:
```diff
@@ -0,0 +1,145 @@
+    "Programming Language :: Python :: 3.12",
+    "Operating System :: OS Independent"
+]
+dependencies = [
```
I'm unsure about this one, for two reasons: first, we don't actually know which specific versions we need. Second, if we choose some sufficiently recent versions (and check that everything is working), we artificially restrict their versions which might not be necessary.

adasilva replied:
I can live with it.

cgahr commented:
> Profile saving is not working for me. I can look into it next week.

I just had a similiar issue, upon restarting however, I could not reproduce the issue.

Is it possible that your selection disappeared? Afaik, without a selection, you cannot save the profile. Or should we save all profiles upon clicking `save profile`?

afonsoguerra commented:
Dear @cgahr, just bumped into this and you saved me tons of time! I can confirm that profiles are not working, other than that it is great. Many many thanks for your rewrite, and obviously to @adasilva as well for the original!

cgahr commented:
Happy to hear that everything (but profiles) is working for. 

I encounter the same bug, but tbh I have no idea how to fix it...

afonsoguerra commented:
I have it all working now. I'll try to add it here. It is probably a bit too verbose with the print statements/debug messages that, like breadcrumbs, I left along the way, but it does work :) I also changed the crop lines to red. while not my favourite colour, the default on my system was white which wasn't very readable, feel free to pick another colour for the final release, but make sure you specify something and don't leave the default. All in all, I think this is going to be super useful! I might expand this to allow setting different output resolutions to better match different devices, but not essential for me right now.

cgahr commented:
Nice! I think the easiest way to proceed would then be merging this commit first and creating a new PR fixing the profile bug.

What do you think?

afonsoguerra commented:
I couldn't quite pull that off (pun intended), but I managed to do a separate PR that integrates all your changes from here... not ideal but I think it should still provide fair attribution and give an opportunity for @adasilva to comment. Have a go and see if I missed anything :) always happy to learn.

cgahr commented:
@afonsoguerra I'll address your PR in the pull request.

I added some instructions for people wanting to help develop or fix bugs. In addition, I added a github action to run the pre-commit hooks on automatically on pull request, to enforce a consistent code style.

@adasilva If everything (but the profile bug) looks fine to you, I propose to merge this PR. Afterwards, @afonsoguerra can create a new PR to fix the profile bug. What do you think?

adasilva commented on pyproject.toml:
```diff
@@ -142,4 +200,5 @@ spaces_indent_inline_array = 4
 
 [tool.tomlsort.overrides]
 "project.classifiers".inline_arrays = false
+"tool.pytest.ini_options.filterwarnings".inline_arrays = false
```
I'm not familiar with this option. Can you provide some documentation? Also, there are not currently any tests, and pytest is not installed. Is this being used at all?

cgahr replied:
tomlsort sorts toml files. However, we don't want the key `tool.pytest.ini_options.filterwarnings` to be sorted since there, the ordering is important.

Since we don't have any test, I can delete that.

adasilva commented on .github/workflows/main.yml:
Please add a commit hash to all actions since tags are mutable, but commit hashes are not. The tag can be placed in a comment for human-readability.

adasilva commented on .pre-commit-config.yaml:
```diff
@@ -1,26 +1,53 @@
+    hooks:
+      - id: mypy
         exclude: tests
+        # - repo: local
```
Remove comments.

adasilva commented:
> @afonsoguerra I'll address your PR in the pull request.
> 
> I added some instructions for people wanting to help develop or fix bugs. In addition, I added a github action to run the pre-commit hooks on automatically on pull request, to enforce a consistent code style.
> 
> @adasilva If everything (but the profile bug) looks fine to you, I propose to merge this PR. Afterwards, @afonsoguerra can create a new PR to fix the profile bug. What do you think?

Agree - @afonsoguerra can create a new PR with the bugfix after merging this one. It will make it easier to review the bug fix.

adasilva reviewed (changes requested):
A couple cleanup comments. 

In the future, it would be better to wait to make configuration changes until the original pull request has been merged, it took me longer to get to review of the new stuff than the changes I had already reviewed.

cgahr commented:
> A couple cleanup comments.
> 
> In the future, it would be better to wait to make configuration changes until the original pull request has been merged, it took me longer to get to review of the new stuff than the changes I had already reviewed.

I'm sorry about that. I intended to add some info for future contributers but mixed multiple things together. Do you want me to revert these commits?

adasilva reviewed (approved):
No need to revert, all changes look good!
